pkg/dockerregistry/server/auth.go

```diff
@@ -272,6 +273,59 @@ func verifyOpenShiftUser(ctx context.Context, client client.UsersInterface) erro
 	return nil
 }
 
+// authorizeRepositoryAccess returns nil if the given client is granted access to the given
+// <namespace>/<imageRepo> repository. Otherwise the access is denied.
+func authorizeRepositoryAccess(ctx context.Context, namespace, imageRepo, verb string, client client.LocalSubjectAccessReviewsNamespacer) error {
+	if isCrossRepoMountFromSourceRepository(ctx, namespace, imageRepo, verb) {
+		return nil
+	}
+
+	return verifyImageStreamAccess(ctx, namespace, imageRepo, verb, client)
+}
+
+// isCrossRepoMountFromSourceRepository handles a case of accessing source repository requested by a cross
+// repository mount during a blob upload. During a push with the cross-repo mount, there are two authorization
+// checks:
+//
+//   1. repository:<source-namespace>/<source-name>:pull
+//   2. repository:<target-namespace>/<target-name>:push
+//
+// This function always grants access in the first case because OpenShift registry treats all the local blobs
+// as accessible once authorized to access given repository (target one).
+func isCrossRepoMountFromSourceRepository(ctx context.Context, namespace, imageRepo, verb string) bool {
+	if verb != "get" {
+		return false
+	}
+	if method := strings.ToLower(getStringValueFromContext(ctx, "http.request.method", "")); method != "post" {
+		return false
+	}
+
+	uriString := getStringValue(ctx, "http.request.uri")
+	uri, err := url.Parse(uriString)
+	if err != nil {
+		context.GetLogger(ctx).Warnf("failed to parse requested uri %q: %v", uri, err)
+		return false
+	}
+	from, exists := uri.Query()["from"]
```

Comment:
it looks like the verifyImageStreamAccess check is skipped if this function returns true, so I have two main questions:
1. where is the access check done prior to this function being called?
2. can the `from` value (or the uri it comes from) be spoofed, allowing the verifyImageStreamAccess check to be skipped?